src/uu/basenc/BENCHMARKING.md

```diff
@@ -0,0 +1,177 @@
+```
+
+#### "--base16", no line wrapping
+
+➖ Slower than GNU Core Utilities
+
+```
+❯ poop \
+    '/usr/bin/basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte' \
+    './target/release/basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte'
+
+Benchmark 1 (6 runs): /usr/bin/basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte
+  measurement          mean ± σ            min … max           outliers         delta
+  wall_time           836ms ± 13.3ms     822ms …  855ms          0 ( 0%)        0%
+  peak_rss           2.05MB ± 73.0KB    1.94MB … 2.12MB          0 ( 0%)        0%
+  cpu_cycles         2.85G  ± 32.8M     2.82G  … 2.91G           0 ( 0%)        0%
+  instructions       14.0G  ± 58.7      14.0G  … 14.0G           0 ( 0%)        0%
+  cache_references   70.0M  ± 6.48M     63.7M  … 78.8M           0 ( 0%)        0%
+  cache_misses        582K  ±  172K      354K  …  771K           0 ( 0%)        0%
+  branch_misses       667K  ± 4.55K      662K  …  674K           0 ( 0%)        0%
+Benchmark 2 (6 runs): ./target/release/basenc --base16 --wrap 0 -- /dev/shm/one-random-gibibyte
+  measurement          mean ± σ            min … max           outliers         delta
+  wall_time           884ms ± 6.38ms     878ms …  895ms          0 ( 0%)        💩+  5.7% ±  1.6%
+  peak_rss           2.65MB ± 66.8KB    2.55MB … 2.74MB          0 ( 0%)        💩+ 29.3% ±  4.4%
+  cpu_cycles         3.15G  ± 8.61M     3.14G  … 3.16G           0 ( 0%)        💩+ 10.6% ±  1.1%
+  instructions       10.5G  ±  275      10.5G  … 10.5G           0 ( 0%)        ⚡- 24.9% ±  0.0%
+  cache_references   93.5M  ± 6.10M     87.2M  …  104M           0 ( 0%)        💩+ 33.7% ± 11.6%
+  cache_misses        415K  ± 52.3K      363K  …  474K           0 ( 0%)          - 28.8% ± 28.0%
+  branch_misses      1.43M  ± 4.82K     1.42M  … 1.43M           0 ( 0%)        💩+113.9% ±  0.9%
+```
+
+### Decoding performance
+
+#### "--base32hex"
+
+➕ Faster than GNU Core Utilities
+
+```
+❯ hyperfine \
+    --sort \
+    command \
+    -- \
+    '/usr/bin/basenc --base32hex --decode -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null' \
+    './target/release/basenc --base32hex --decode -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null'
+
+Benchmark 1: /usr/bin/basenc --base32hex --decode -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
+  Time (mean ± σ):      7.154 s ±  0.082 s    [User: 6.802 s, System: 0.323 s]
+  Range (min … max):    7.051 s …  7.297 s    10 runs
+
+Benchmark 2: ./target/release/basenc --base32hex --decode -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
+  Time (mean ± σ):      2.679 s ±  0.025 s    [User: 2.446 s, System: 0.221 s]
+  Range (min … max):    2.649 s …  2.718 s    10 runs
+
+Relative speed comparison
+        2.67 ±  0.04  /usr/bin/basenc --base32hex --decode -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
```

Comment:
```
        3.27 ±  0.57  /usr/bin/basenc --base32hex --decode -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
        1.00          ./target/release/coreutils basenc --base32hex --decode -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
        1.04 ±  0.17  ./coreutils basenc --base32hex --decode -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
```

getting something similar here :)

Reply:
Yep, looks like in some scenarios the new implementation isn't much faster. The main advantage here would just be the constant memory usage, and faster handling of "--ignore-garbage".

```
❯ hyperfine \
    --sort \
    command \
    -- \
    '/usr/bin/basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null' \
    'coreutils basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null' \
    './target/release/basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null'

Benchmark 1: /usr/bin/basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
  Time (mean ± σ):     10.450 s ±  0.201 s    [User: 10.060 s, System: 0.327 s]
  Range (min … max):   10.242 s … 10.876 s    10 runs
 
Benchmark 2: coreutils basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
  Time (mean ± σ):     26.267 s ±  0.225 s    [User: 24.536 s, System: 1.566 s]
  Range (min … max):   25.986 s … 26.675 s    10 runs
 
Benchmark 3: ./target/release/basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
  Time (mean ± σ):      2.703 s ±  0.019 s    [User: 2.485 s, System: 0.201 s]
  Range (min … max):    2.664 s …  2.725 s    10 runs
 
Relative speed comparison
        3.87 ±  0.08  /usr/bin/basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
        9.72 ±  0.11  coreutils basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
        1.00          ./target/release/basenc --base32hex --decode --ignore-garbage -- /dev/shm/one-random-gibibyte-base32hex-encoded 1>/dev/null
```